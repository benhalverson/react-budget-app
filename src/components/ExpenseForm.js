import React, { useState } from 'react';
import moment from 'moment';
import { SingleDatePicker } from 'react-dates';
import 'react-dates/initialize';

function useInput({ initialValue, name }) {
  const [value, setValue] = useState(initialValue);
  function onChange(e) {
    const updated = e.target.value;
    setValue(updated);
  }
  return [value, onChange];
}

export default function ExpenseForm({ expense, onSubmit }) {
  const [description, onDescriptionChange] = useInput(
    expense ? expense.description : ''
  );
  const [note, onNoteChange] = useInput(expense ? expense.note : '');
  const [category, onCategoryChange] = useInput(
    expense ? expense.category : ''
  );
  const [amount, setAmount] = useState(
    expense ? (expense.amount / 100).toString() : ''
  );
  const [createdAt, setCreatedAt] = useState(
    expense ? moment(expense.createdAt) : moment()
  );
  const [calendarFocused, setCalendarFocused] = useState(false);
  const [error, setError] = useState('');

  function onAmountChange(e) {
    const amount = e.target.value;

    if (!amount || amount.match(/^\d{1,}(\.\d{0,2})?$/)) {
      setAmount(amount);
    }
  }
  function onDateChange(createdAt) {
    if (createdAt) {
      setCreatedAt(createdAt);
    }
  }
  function onFocusChange({ focused }) {
    setCalendarFocused(focused);
  }
  function onFormSubmit(e) {
    e.preventDefault();

    if (!description || !amount) {
      setError('Please provide description, category, and amount.');
    } else {
      setError('');
      onSubmit({
        description,
        amount: parseFloat(amount) * 100,
        category,
        createdAt: createdAt.valueOf(),
        note,
      });
    }
    console.log(category);
  }

  return (
    <form className='form' onSubmit={onFormSubmit}>
      {error && <p className='form__error'>{error}</p>}
      <input
        type='text'
        placeholder='Description'
        autoFocus
        className='text-input'
        value={description}
        onChange={onDescriptionChange}
      />
      <input
        type='text'
        placeholder='Category'
        autoFocus
        className='text-input'
        value={category}
        onChange={onCategoryChange}
      />
      <input
        type='text'
        placeholder='Amount'
        className='text-input'
        value={amount}
        onChange={onAmountChange}
      />
      <SingleDatePicker
        date={createdAt}
        onDateChange={onDateChange}
        focused={calendarFocused}
        onFocusChange={onFocusChange}
        numberOfMonths={1}
        isOutsideRange={() => false}
      />
      <textarea
        placeholder='Add a note for your expense (optional)'
        className='textarea'
        value={note}
        onChange={onNoteChange}
      />
      <div>
        <button className='button'>Save Expense</button>
      </div>
    </form>
  );
}
