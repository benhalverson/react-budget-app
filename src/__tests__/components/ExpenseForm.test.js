import React from 'react';
import { shallow } from 'enzyme';
import moment from 'moment';
import ExpenseForm from '../../components/ExpenseForm';
const expenses = [
  {
    id: '1',
    description: 'Gum',
    note: '',
    amount: 195,
    category: 'Food',
    createdAt: 0
  },
  {
    id: '2',
    description: 'Rent',
    note: '',
    amount: 109500,
    category: 'bill',
    createdAt: moment(0)
      .subtract(4, 'days')
      .valueOf()
  },
  {
    id: '3',
    description: 'Credit Card',
    note: '',
    amount: 4500,
    category: 'bill',
    createdAt: moment(0)
      .add(4, 'days')
      .valueOf()
  }
];

test('should render ExpenseForm correctly', () => {
  const wrapper = shallow(<ExpenseForm />);
  expect(wrapper).toMatchSnapshot();
});

test('should render ExpenseForm correctly with expense data', () => {
  const wrapper = shallow(<ExpenseForm expense={expenses[1]} />);
  expect(wrapper).toMatchSnapshot();
});

test('should render error for invalid form submission', () => {
  const wrapper = shallow(<ExpenseForm />);
  expect(wrapper).toMatchSnapshot();
  wrapper.find('form').simulate('submit', {
    preventDefault: () => {}
  });
  expect(wrapper.state('error').length).toBeGreaterThan(0);
  expect(wrapper).toMatchSnapshot();
});

test('should set description on input change', () => {
  const value = 'New description';
  const wrapper = shallow(<ExpenseForm />);
  wrapper
    .find('input')
    .at(0)
    .simulate('change', {
      target: { value }
    });
  console.log('description', value);
  console.log(wrapper.value);
  console.log('end of description');
  expect(wrapper.state('description')).toBe(value);
});

test('should set category on input change', () => {
  const value = 'bills';
  const wrapper = shallow(<ExpenseForm />);
  wrapper
    .find('input')
    .at(0)
    .simulate('change', {
      target: { value }
    });
  expect(wrapper.state('category')).toBe(value);
});

test('should set note on textarea change', () => {
  const value = 'New note value';
  const wrapper = shallow(<ExpenseForm />);
  wrapper.find('textarea').simulate('change', {
    target: { value }
  });
  expect(wrapper.state('note')).toBe(value);
});

test('should set amount if valid input', () => {
  const value = '23.50';
  const wrapper = shallow(<ExpenseForm />);
  wrapper
    .find('input')
    .at(1)
    .simulate('change', {
      target: { value }
    });
  expect(wrapper.state('amount')).toBe(value);
});

test('should not set amount if invalid input', () => {
  const value = '12.122';
  const wrapper = shallow(<ExpenseForm />);
  wrapper
    .find('input')
    .at(1)
    .simulate('change', {
      target: { value }
    });
  expect(wrapper.state('amount')).toBe('');
});

test('should call onSubmit prop for valid form submission', () => {
  const onSubmitSpy = jest.fn();
  const wrapper = shallow(
    <ExpenseForm expense={expenses[0]} onSubmit={onSubmitSpy} />
  );
  wrapper.find('form').simulate('submit', {
    preventDefault: () => {}
  });
  expect(wrapper.state('error')).toBe('');
  expect(onSubmitSpy).toHaveBeenLastCalledWith({
    description: expenses[0].description,
    amount: expenses[0].amount,
    note: expenses[0].note,
    createdAt: expenses[0].createdAt
  });
});

test('should set new date on date change', () => {
  const now = moment();
  const wrapper = shallow(<ExpenseForm />);
  wrapper.find('withStyles(SingleDatePicker)').prop('onDateChange')(now);
  expect(wrapper.state('createdAt')).toEqual(now);
});

test('should set calendar focus on change', () => {
  const focused = true;
  const wrapper = shallow(<ExpenseForm />);
  wrapper.find('withStyles(SingleDatePicker)').prop('onFocusChange')({
    focused
  });
  expect(wrapper.state('calendarFocused')).toBe(focused);
});
