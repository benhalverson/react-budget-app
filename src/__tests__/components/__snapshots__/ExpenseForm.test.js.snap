// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render ExpenseForm correctly 1`] = `ShallowWrapper {}`;

exports[`should render ExpenseForm correctly with expense data 1`] = `ShallowWrapper {}`;

exports[`should render error for invalid form submission 1`] = `ShallowWrapper {}`;

exports[`should render error for invalid form submission 2`] = `ShallowWrapper {}`;
